File: wrench-app/src/main/java/com/izettle/wrench/configurationlist/ConfigurationsFragment.java
package com.izettle.wrench.configurationlist;

import android.app.ActivityManager;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.SearchView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ViewAnimator;

import com.izettle.wrench.R;
import com.izettle.wrench.core.Bolt;
import com.izettle.wrench.database.WrenchApplication;
import com.izettle.wrench.database.WrenchConfigurationWithValues;
import com.izettle.wrench.databinding.FragmentConfigurationsBinding;
import com.izettle.wrench.di.Injectable;
import com.izettle.wrench.dialogs.booleanvalue.BooleanValueFragment;
import com.izettle.wrench.dialogs.booleanvalue.BooleanValueFragmentArgs;
import com.izettle.wrench.dialogs.enumvalue.EnumValueFragment;
import com.izettle.wrench.dialogs.enumvalue.EnumValueFragmentArgs;
import com.izettle.wrench.dialogs.integervalue.IntegerValueFragment;
import com.izettle.wrench.dialogs.integervalue.IntegerValueFragmentArgs;
import com.izettle.wrench.dialogs.scope.ScopeFragment;
import com.izettle.wrench.dialogs.stringvalue.StringValueFragment;
import com.izettle.wrench.dialogs.stringvalue.StringValueFragmentArgs;

import java.util.List;

import javax.inject.Inject;

public class ConfigurationsFragment extends Fragment implements SearchView.OnQueryTextListener, ConfigurationRecyclerViewAdapter.Listener, Injectable {
    private static final String STATE_FILTER = "STATE_FILTER";
    @Inject
    ViewModelProvider.Factory viewModelFactory;
    private FragmentConfigurationsBinding fragmentConfigurationsBinding;
    private CharSequence currentFilter;
    private SearchView searchView;
    private ConfigurationViewModel model;


    public ConfigurationsFragment() {
    }

    public static ConfigurationsFragment newInstance() {
        ConfigurationsFragment fragment = new ConfigurationsFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private void updateConfigurations(List<WrenchConfigurationWithValues> wrenchConfigurations) {
        ConfigurationRecyclerViewAdapter adapter = (ConfigurationRecyclerViewAdapter) fragmentConfigurationsBinding.list.getAdapter();
        if (adapter == null) {
            adapter = new ConfigurationRecyclerViewAdapter(this, model);

            fragmentConfigurationsBinding.list.setAdapter(adapter);
        }
        adapter.setItems(wrenchConfigurations);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putCharSequence(STATE_FILTER, searchView.getQuery());
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);

        if (savedInstanceState != null) {
            currentFilter = savedInstanceState.getCharSequence(STATE_FILTER);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        fragmentConfigurationsBinding = FragmentConfigurationsBinding.inflate(inflater, container, false);

        return fragmentConfigurationsBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        assert getArguments() != null;

        model = ViewModelProviders.of(this, viewModelFactory).get(ConfigurationViewModel.class);
        ConfigurationsFragmentArgs args = ConfigurationsFragmentArgs.fromBundle(getArguments());
        model.setApplicationId(args.getApplicationId());

        fragmentConfigurationsBinding.list.setLayoutManager(new LinearLayoutManager(getContext()));

        model.getWrenchApplication().observe(this, wrenchApplication -> {
            if (wrenchApplication != null) {
                ((AppCompatActivity) getActivity()).getSupportActionBar().setTitle(wrenchApplication.applicationLabel());
            }
        });

        model.getDefaultScopeLiveData().observe(this, scope -> {
            if (scope != null && fragmentConfigurationsBinding.list.getAdapter() != null) {
                fragmentConfigurationsBinding.list.getAdapter().notifyDataSetChanged();
            }
        });

        model.getSelectedScopeLiveData().observe(this, scope -> {
            if (scope != null && fragmentConfigurationsBinding.list.getAdapter() != null) {
                fragmentConfigurationsBinding.list.getAdapter().notifyDataSetChanged();
            }
            fragmentConfigurationsBinding.scopeButton.setText(scope.getName());
        });

        fragmentConfigurationsBinding.scopeButton.setOnClickListener(v -> ScopeFragment.newInstance(args.getApplicationId()).show(getChildFragmentManager(), null));

        model.getConfigurations().observe(this, wrenchConfigurationWithValues -> {
            if (wrenchConfigurationWithValues != null) {
                updateConfigurations(wrenchConfigurationWithValues);
            }
        });

        model.isListEmpty().observe(this, isEmpty -> {
            ViewAnimator animator = fragmentConfigurationsBinding.animator;
            if (isEmpty == null || isEmpty) {
                animator.setDisplayedChild(animator.indexOfChild(fragmentConfigurationsBinding.noConfigurationsEmptyView));
            } else {
                animator.setDisplayedChild(animator.indexOfChild(fragmentConfigurationsBinding.listWrapper));
            }
        });
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_configurations_list, menu);

        MenuItem item = menu.findItem(R.id.action_filter_configurations);
        searchView = (SearchView) item.getActionView();
        searchView.setOnQueryTextListener(this);

        item.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
            @Override
            public boolean onMenuItemActionCollapse(MenuItem item) {
                return true; // Return true to collapse action view
            }

            @Override
            public boolean onMenuItemActionExpand(MenuItem item) {
                return true; // Return true to expand action view
            }
        });

        if (!TextUtils.isEmpty(currentFilter)) {
            item.expandActionView();
            searchView.setQuery(currentFilter, true);
        }
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        model.setQuery(newText);
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_restart_application: {
                model.getWrenchApplication().observe(this, new Observer<WrenchApplication>() {
                    @Override
                    public void onChanged(@Nullable WrenchApplication wrenchApplication) {
                        model.getWrenchApplication().removeObserver(this);

                        if (wrenchApplication == null) {
                            return;
                        }

                        ActivityManager activityManager = (ActivityManager) getContext().getSystemService(Context.ACTIVITY_SERVICE);
                        activityManager.killBackgroundProcesses(wrenchApplication.packageName());

                        Intent intent = getContext().getPackageManager().getLaunchIntentForPackage(wrenchApplication.packageName());
                        if (intent != null) {
                            ConfigurationsFragment.this.getContext().startActivity(Intent.makeRestartActivityTask(intent.getComponent()));
                        } else if (ConfigurationsFragment.this.getView() != null) {
                            Snackbar.make(ConfigurationsFragment.this.getView(), R.string.application_not_installed, Snackbar.LENGTH_LONG).show();
                        }
                    }
                });

                return true;
            }
            case R.id.action_application_settings: {
                model.getWrenchApplication().observe(this, new Observer<WrenchApplication>() {
                    @Override
                    public void onChanged(@Nullable WrenchApplication wrenchApplication) {
                        model.getWrenchApplication().removeObserver(this);
                        if (wrenchApplication == null) {
                            return;
                        }

                        startActivity(new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.fromParts("package", wrenchApplication.packageName(), null)));
                    }
                });
                return true;
            }
            case R.id.action_delete_application: {
                AsyncTask.execute(() -> model.deleteApplication(model.getWrenchApplication().getValue()));
                getActivity().finish();
                return true;
            }
            default: {
                return super.onOptionsItemSelected(item);
            }
        }
    }

    @Override
    public void configurationClicked(View v, WrenchConfigurationWithValues configuration) {
        if (model.getSelectedScopeLiveData().getValue() == null) {
            Snackbar.make(fragmentConfigurationsBinding.getRoot(), "No selected scope found", Snackbar.LENGTH_LONG).show();
            return;
        }

        int selectedScopeId = (int) model.getSelectedScopeLiveData().getValue().id();

        if (TextUtils.equals(String.class.getName(), configuration.getType()) ||
                TextUtils.equals(Bolt.TYPE.STRING, configuration.getType())) {

            StringValueFragment.newInstance(new StringValueFragmentArgs.Builder((int) configuration.getId(), selectedScopeId).build()).show(getChildFragmentManager(), null);

        } else if (TextUtils.equals(Integer.class.getName(), configuration.getType()) ||
                TextUtils.equals(Bolt.TYPE.INTEGER, configuration.getType())) {

            IntegerValueFragment.newInstance(new IntegerValueFragmentArgs.Builder((int) configuration.getId(), selectedScopeId).build()).show(getChildFragmentManager(), null);

        } else if (TextUtils.equals(Boolean.class.getName(), configuration.getType()) ||
                TextUtils.equals(Bolt.TYPE.BOOLEAN, configuration.getType())) {

            BooleanValueFragment.newInstance(new BooleanValueFragmentArgs.Builder((int) configuration.getId(), selectedScopeId).build()).show(getChildFragmentManager(), null);

        } else if (TextUtils.equals(Enum.class.getName(), configuration.getType()) ||
                TextUtils.equals(Bolt.TYPE.ENUM, configuration.getType())) {

            EnumValueFragment.newInstance(new EnumValueFragmentArgs.Builder((int) configuration.getId(), selectedScopeId).build()).show(getChildFragmentManager(), null);

        } else {

            Snackbar.make(fragmentConfigurationsBinding.getRoot(), "Not sure what to do with type: " + configuration.getType(), Snackbar.LENGTH_LONG).show();
        }
    }
}

File: wrench-app/src/main/java/com/izettle/wrench/configurationlist/ConfigurationsFragment.kt
package com.izettle.wrench.configurationlist

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import com.izettle.wrench.R
import com.izettle.wrench.core.Bolt
import com.izettle.wrench.database.WrenchApplication
import com.izettle.wrench.database.WrenchConfigurationWithValues
import com.izettle.wrench.databinding.FragmentConfigurationsBinding
import com.izettle.wrench.dialogs.booleanvalue.BooleanValueFragment
import com.izettle.wrench.dialogs.booleanvalue.BooleanValueFragmentArgs
import com.izettle.wrench.dialogs.enumvalue.EnumValueFragment
import com.izettle.wrench.dialogs.enumvalue.EnumValueFragmentArgs
import com.izettle.wrench.dialogs.integervalue.IntegerValueFragment
import com.izettle.wrench.dialogs.integervalue.IntegerValueFragmentArgs
import com.izettle.wrench.dialogs.scope.ScopeFragment
import com.izettle.wrench.dialogs.stringvalue.StringValueFragment
import com.izettle.wrench.dialogs.stringvalue.StringValueFragmentArgs
import org.koin.androidx.viewmodel.ext.viewModel


class ConfigurationsFragment : Fragment(), SearchView.OnQueryTextListener, ConfigurationRecyclerViewAdapter.Listener {
    private lateinit var fragmentConfigurationsBinding: FragmentConfigurationsBinding
    private var currentFilter: CharSequence? = null
    private var searchView: SearchView? = null
    private val model: ConfigurationViewModel by viewModel()

    private fun updateConfigurations(wrenchConfigurations: List<WrenchConfigurationWithValues>) {
        var adapter = fragmentConfigurationsBinding.list.adapter as ConfigurationRecyclerViewAdapter?
        if (adapter == null) {
            adapter = ConfigurationRecyclerViewAdapter(this, model)

            fragmentConfigurationsBinding.list.adapter = adapter
        }
        adapter.submitList(wrenchConfigurations)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putCharSequence(STATE_FILTER, searchView!!.query)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)

        if (savedInstanceState != null) {
            currentFilter = savedInstanceState.getCharSequence(STATE_FILTER)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return FragmentConfigurationsBinding.inflate(inflater, container, false).also { fragmentConfigurationsBinding = it }.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        assert(arguments != null)

        val args = ConfigurationsFragmentArgs.fromBundle(arguments!!)
        model.setApplicationId(args.applicationId)

        fragmentConfigurationsBinding.list.layoutManager = LinearLayoutManager(context)

        model.wrenchApplication.observe(this, Observer { wrenchApplication ->
            if (wrenchApplication != null) {
                (activity as AppCompatActivity).supportActionBar!!.title = wrenchApplication.applicationLabel
            }
        })

        model.defaultScopeLiveData.observe(this, Observer { scope ->
            if (scope != null && fragmentConfigurationsBinding.list.adapter != null) {
                fragmentConfigurationsBinding.list.adapter!!.notifyDataSetChanged()
            }
        })

        model.selectedScopeLiveData.observe(this, Observer { scope ->
            if (scope != null && fragmentConfigurationsBinding.list.adapter != null) {
                fragmentConfigurationsBinding.list.adapter!!.notifyDataSetChanged()
            }
            // fragmentConfigurationsBinding.scopeButton.text = scope!!.name
        })

        model.configurations.observe(this, Observer { wrenchConfigurationWithValues ->
            if (wrenchConfigurationWithValues != null) {
                updateConfigurations(wrenchConfigurationWithValues)
            }
        })

        model.isListEmpty.observe(this, Observer { isEmpty ->
            val animator = fragmentConfigurationsBinding.animator
            if (isEmpty == null || isEmpty) {
                animator.displayedChild = animator.indexOfChild(fragmentConfigurationsBinding.noConfigurationsEmptyView)
            } else {
                animator.displayedChild = animator.indexOfChild(fragmentConfigurationsBinding.list)
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_configurations_list, menu)

        val item = menu.findItem(R.id.action_filter_configurations)
        searchView = item.actionView as SearchView
        searchView!!.setOnQueryTextListener(this)

        item.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                return true // Return true to collapse action view
            }

            override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                return true // Return true to expand action view
            }
        })

        if (currentFilter.isNullOrBlank()) {
            item.expandActionView()
            searchView!!.setQuery(currentFilter, true)
        }
    }

    override fun onQueryTextChange(newText: String): Boolean {
        model.setQuery(newText)
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_restart_application -> {
                model.wrenchApplication.observe(this, object : Observer<WrenchApplication> {
                    override fun onChanged(wrenchApplication: WrenchApplication?) {
                        model.wrenchApplication.removeObserver(this)

                        if (wrenchApplication == null) {
                            return
                        }

                        val activityManager = context!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
                        activityManager.killBackgroundProcesses(wrenchApplication.packageName)

                        val intent = context!!.packageManager.getLaunchIntentForPackage(wrenchApplication.packageName)
                        if (intent != null) {
                            context!!.startActivity(Intent.makeRestartActivityTask(intent.component))
                        } else if (this@ConfigurationsFragment.view != null) {
                            Snackbar.make(this@ConfigurationsFragment.view!!, R.string.application_not_installed, Snackbar.LENGTH_LONG).show()
                        }
                    }
                })

                return true
            }
            R.id.action_application_settings -> {
                model.wrenchApplication.observe(this, object : Observer<WrenchApplication> {
                    override fun onChanged(wrenchApplication: WrenchApplication?) {
                        model.wrenchApplication.removeObserver(this)
                        if (wrenchApplication == null) {
                            return
                        }

                        startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.fromParts("package", wrenchApplication.packageName, null)))
                    }
                })
                return true
            }
            R.id.action_delete_application -> {
                model.deleteApplication(model.wrenchApplication.value!!)
                Navigation.findNavController(fragmentConfigurationsBinding.root).navigateUp()
                return true
            }
            R.id.action_change_scope -> {
                val args = ConfigurationsFragmentArgs.fromBundle(arguments!!)
                ScopeFragment.newInstance(args.applicationId).show(childFragmentManager, null)
                return true
            }
            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
    }

    override fun configurationClicked(v: View, configuration: WrenchConfigurationWithValues) {
        if (model.selectedScopeLiveData.value == null) {
            Snackbar.make(fragmentConfigurationsBinding.root, "No selected scope found", Snackbar.LENGTH_LONG).show()
            return
        }

        val selectedScopeId = model.selectedScopeLiveData.value!!.id

        if (TextUtils.equals(String::class.java.name, configuration.type) || TextUtils.equals(Bolt.TYPE.STRING, configuration.type)) {

            StringValueFragment.newInstance(StringValueFragmentArgs.Builder(configuration.id, selectedScopeId).build()).show(childFragmentManager, null)

        } else if (TextUtils.equals(Int::class.java.name, configuration.type) || TextUtils.equals(Bolt.TYPE.INTEGER, configuration.type)) {

            IntegerValueFragment.newInstance(IntegerValueFragmentArgs.Builder(configuration.id, selectedScopeId).build()).show(childFragmentManager, null)

        } else if (TextUtils.equals(Boolean::class.java.name, configuration.type) || TextUtils.equals(Bolt.TYPE.BOOLEAN, configuration.type)) {

            BooleanValueFragment.newInstance(BooleanValueFragmentArgs.Builder(configuration.id, selectedScopeId).build()).show(childFragmentManager, null)

        } else if (TextUtils.equals(Enum::class.java.name, configuration.type) || TextUtils.equals(Bolt.TYPE.ENUM, configuration.type)) {

            EnumValueFragment.newInstance(EnumValueFragmentArgs.Builder(configuration.id, selectedScopeId).build()).show(childFragmentManager, null)

        } else {

            Snackbar.make(fragmentConfigurationsBinding.root, "Not sure what to do with type: " + configuration.type!!, Snackbar.LENGTH_LONG).show()
        }
    }

    companion object {
        private const val STATE_FILTER = "STATE_FILTER"
    }
}
